rust/opendp/src/lib.rs

```diff
@@ -38,26 +38,31 @@
 //! use opendp::core;
 //! use opendp::meas;
 //! use opendp::trans;
+//! use opendp::trans::{MakeTransformation0, MakeTransformation1, MakeTransformation2, MakeTransformation3};
+//! use opendp::dist::{HammingDistance, L1Sensitivity};
+//! use opendp::core::{ChainTT, ChainMT};
+//! use opendp::meas::{MakeMeasurement2, LaplaceMechanism, MakeMeasurement1};
+//!
 //! pub fn example() {
 //!     let data = "56\n15\n97\n56\n6\n17\n2\n19\n16\n50".to_owned();
 //!     let bounds = (0.0, 100.0);
 //!     let epsilon = 1.0;
 //!     let sigma = (bounds.1 - bounds.0) / epsilon;
 //!
 //!     // Construct a Transformation to load the numbers.
-//!     let split_lines = trans::make_split_lines();
-//!     let parse_series = trans::make_parse_series::<f64>(true);
-//!     let load_numbers = core::make_chain_tt(&parse_series, &split_lines);
+//!     let split_lines = trans::SplitLines::<HammingDistance>::make();
+//!     let parse_series = trans::ParseSeries::<f64, HammingDistance>::make(true);
+//!     let load_numbers = ChainTT::make(&parse_series, &split_lines);
 //!
-//!     // Construct a Measurment to calculate a noisy sum.
-//!     let clamp = trans::make_clamp(bounds.0, bounds.1);
-//!     let bounded_sum = trans::make_bounded_sum_l1(bounds.0, bounds.1);
-//!     let laplace = meas::make_base_laplace(sigma);
-//!     let intermediate = core::make_chain_tt(&bounded_sum, &clamp);
-//!     let noisy_sum = core::make_chain_mt(&laplace, &intermediate);
+//!     // Construct a Measurement to calculate a noisy sum.
+//!     let clamp = trans::Clamp::make(bounds.0, bounds.1);
+//!     let bounded_sum = trans::BoundedSum::make2(bounds.0, bounds.1);
```

Comment:
Don't miss that this is now `::make2` to disambiguate which trait to dispatch to. `BoundedSum::make` is implemented for both 2 and 3 arguments.

Reply:
If you impl more than one trait with the same function (ex: `MakeTransformation0`, `MakeTransformation1`), rust can't determine which trait to dispatch to based on type or arity.
An example of this is the impls for BoundedSum. The restricted sensitivity impl takes an extra argument for dataset length, causing calling sites to use the fully qualified syntax to disambiguate:
```let bounded_sum = <trans::BoundedSum<_, _, _> as MakeTransformation2<_, _, _, _, _, _>>::make(bounds.0, bounds.1);```
One way we could disambiguate easier is by changing the function from `make` to `makeN`. In most cases, you can still automatically dispatch to the right impl, so the constructor traits can define a default implementation of `make` that points to `makeN`. This way, if you do run into issues with ambiguous trait impls, you can add the arity after `make`, instead of using the fully qualified syntax. This does mean adding a second impl with different arity can be a breaking change to make if both traits happen to be imported.

Reply:
Oh interesting. I remember reading about this issue in the Rust docs, but haven't encountered it in code before now.

The way you split these out into separate functions with explicit arity is a cool solution.